### ISSUE TITLE
[Bug] PD returns empty completions under queue depth; degraded rows share KV slots with live requests

### ISSUE THREAD

**Issue by jzheng17** [sgl-project/sglang-omni#1704]
A correctness defect in the PD path, with the measurement that localizes it. It is not a performance issue and it is silent.

## What a user sees

A request returns HTTP 200, `finish_reason: "stop"`, `completion_tokens: 1`, correct `prompt_tokens`, and an empty delta. Two SSE frames, the terminal chunk and `[DONE]`, with no content frame. The single token is 151645, `<|im_end|>`, and it ends the request through `FINISH_MATCHED_TOKEN` as a matched stop token rather than as a member of `eos_token_ids` — so a probe that checks `eos_token_ids` cannot see this population, which cost us one instrumented run.

The response is well formed. Nothing logs an error.

## Rate

PD only. Measured on a clean two-GPU pair with mixed prompt lengths, `mem_fraction_static=0.87`, zero transport errors:

| offered | achieved | in-flight p50 | empty completions per 3000 |
| --- | --- | --- | --- |
| 6 | 5.96 | 8 | 0 |
| 8 | 7.92 | 12 | 0 |
| 10 | 9.90 | 20 | 1 |
| 12 | 10.64 | 175 | 24 |
| 16 | 10.91 | 535 | 48 |

A colocated replica under an identical driver produced 0 in 3000, and 0 in 28,472 at the longest prompt length.

## What it is not

**Not the server args PD forces.** A colocated arm with `page_size=1` and `disable_radix_cache` set, same driver, same mixed lengths, same long-lived server, produced zero.

**Not the KV transfer or the handoff.** The single token is sampled by the prefill half before any transfer occurs, so a decode-side probe never fires for this population.

**Not sampling numerics.** Decoding is greedy. Across two runs the failing rows' logit gaps span the control distribution, with a median of 13 bf16 ULP against the control's 9, and four of eleven above 33 ULP. Colocated never places the terminator in its top two: 0 of 2237 prefill rows, against a rate that predicts about 9.8, Poisson p = 5.6e-5.

**Not prefill batch composition.** Raising `prefill_coalesce_requests` from the default to 8 moved mean batch size 2.73 to 3.11 and cut single-request batches from 20.6% to 6.7%, with no rise in the failure rate. Failures occur at batch size 1.

**Not prompt length.** Short prompts fail once queues are deep. An earlier length correlation was queue depth in disguise.

## What it is

The whole logit vector is depressed, not re-ranked. Failing rows sit 7 to 8 logits below control on both the top and second token while the gap between them is preserved. The control distribution does not move at all across the entire rate ladder — median top logit 22.75 at every rate — so depth does nothing to ordinary rows. It creates a distinct population: zero rows below `l1 = 16` across 5989 rows at the two lowest rates, then 1.4%, 3.4% and 6.3% as depth rises.

That shape is a degraded hidden state at the final position, not a ranking or sampling fault.

**Degraded rows share KV slots with a concurrently allocated request. Control rows never do.**

| | degraded (`l1 <= 16`), n=13 | control, n=61 |
| --- | --- | --- |
| shares a slot with a currently-allocated request | 7 (53.8%) | **0 (0.0%)** |
| median / max shared slots | 475 / 7323 | 0 / **0** |
| own mapping internally unique | 13/13 | 61/61 |

Length-controlled: mid-length 4/6 against 0/21, long 3/7 against 0/20. One captured failure had `n=1963` with 1304 of its slots concurrently mapped by another allocated request, its own mapping clean.

Separating "currently allocated" from "recently seen" was what made this readable. Counting both together gives 86.7% against 3.9%, which looks like a rate difference; counting only live sharers gives an absolute separation.

## Where to look

`_queue_pd_prefill_handoffs` ends with `batch.reqs = retained`, removing handed-off requests from the batch while their KV stays live under `SGLangKVPageLease` until the receiver ACK releases it. That is the only place in the PD path where a request leaves the scheduler's batch while still owning slots, and it is the surface where a slot could reach a new request while the old mapping still points at it. We have not traced the exact ordering that produces the alias, so treat this as where the evidence points rather than as a diagnosis.

## What we did not establish

**A second route exists.** Six of thirteen degraded rows had neither slot sharing nor internal duplication and were still degraded. Whatever explains those is not in the KV map, and finding it needs a hidden-state probe.

**The depth relationship has an unresolved confound.** A fresh server driven straight to the top rate produced 0.20% against the aged server's 1.60% at *higher* in-flight, but it ran on contended cards, so server age and contention are not separated. The 10 to 12 rps knee should not be treated as a clean number.

**Small n.** Thirteen degraded rows in the run that settles the live-sharing question, and the probe gates on `l1 <= 16` plus a 2% sample, so most failures are not captured.

Raw data, probes and per-run analysis are available if useful.

**Comment by jzheng17** [comment]
Replicated on a fresh run, with the prompt-length confound controlled. The association between an empty completion and cross-request KV slot sharing holds.

## What reproduces it

Mixed prompt lengths are required. Uniform long prompts do not reproduce it: six runs at about 7,500 tokens each produced zero empty completions, including one on a tree without any of the placement changes.

    repro_sweep.py --rate 16 --n 3000 --pflags 42,1150,4400 --max-tokens 128

against a two-card PD pair at `mem_fraction_static=0.82`. Two runs gave 0.23% and 0.07% empty, with in-flight p50 of 557 and 549, matching the 535 seen originally.

That fits the mechanism. Mixing lengths makes short requests free small extents while long ones need large ones, which is the condition under which an allocator can hand out a slot that another request still maps. Uniform lengths do not create that pressure. The original data says the same thing from the other side: degraded rows had a median prompt of 7,496 tokens against 196 for controls, so both classes were present in one run.

## The association, within each prompt class

86 rows captured. `xshare` counts this request's KV slots that also appear in another live request's mapping at the same moment.

| Prompt class | Degraded rows | of those, xshare > 0 | Control rows | of those, xshare > 0 |
| --- | --- | --- | --- | --- |
| ~42 tokens | 1 | 0 | 23 | 0 |
| ~1,150-2,000 tokens | 3 | **3** | 25 | **0** |
| ~7,000+ tokens | 7 | **6** | 27 | **1** |

In the two classes where degradation appears at all, that is 9 of 10 degraded rows against 1 of 52 controls.

The length control matters and it strengthens the result rather than explaining it away: degraded rows skew long (median 7,673 against 2,007), so an uncontrolled comparison could have been reporting length rather than sharing. Within a class the difference is still 9/10 against 1/52.

This is the second independent run to show it. The first was 13/15 against 2/51.

## What is not established

How the aliasing arises. `SGLangKVPageLease` keeps prefill KV alive until the receiver's ACK releases the source, so a release path that reclaims a page while a lease still references it would produce exactly this. That is a candidate, not a finding.

Two other things were tested and ruled out. Bounding how many handoffs hold prefill KV at once does not change the rate: across six runs the phenomenon appeared once with the bound and once without, at 0.286% and 0.263%, with identical KV pools on identical cards. And it is not something the placement changes introduced, since a tree without them does not reproduce it either.

**Comment by jzheng17** [comment]
A second signal in the same rows, and it points the two together at one thing: the allocator hands out a slot that is not free.

## Two ways a degraded request's mapping is wrong

`xshare` counts this request's KV slots that another live request also maps. `uniq_ok` is false when the request's own mapping contains the same slot twice, which is a different fault with the same cause.

Within each prompt class, from the 86 rows captured:

| Prompt class | Degraded | xshare > 0 | uniq_ok false | Control | xshare > 0 | uniq_ok false |
| --- | --- | --- | --- | --- | --- | --- |
| ~42 tokens | 1 | 0 | 0 | 23 | 0 | 0 |
| ~1,150-2,000 | 3 | 3 | 2 | 25 | 0 | 0 |
| ~7,000+ | 7 | 6 | 4 | 27 | 1 | 1 |

Taking either fault as the marker: **9 of 11 degraded rows against 2 of 75 controls**.

## The two faults are one fault

A slot that another live request maps and a slot handed twice to the same request are both a slot given out while still in use. The difference is only who else holds it. Read that way the evidence is not about the transfer at all; it is about the bookkeeping that decides a slot is free.

`reuse_gap` supports that. On the rows with duplicate slots it sits at 73 to 78, so those slots had been used very recently, which is what premature reclamation looks like.

Two rows show neither fault, so this does not account for every empty completion. One of them is a 39-token request, the class where degradation is otherwise absent.

## Detail of the degraded rows

| l1 | prompt | slots | unique | xshare | xshare/n | reuse_gap |
| --- | --- | --- | --- | --- | --- | --- |
| 15.81 | 7673 | 7673 | 7673 | 4322 | 0.56 | -1 |
| 13.06 | 7673 | 7673 | 7673 | 3493 | 0.46 | -1 |
| 14.62 | 7673 | 7673 | 7316 | 1135 | 0.15 | 77 |
| 15.38 | 2007 | 2007 | 1535 | 1020 | 0.51 | 73 |
| 15.94 | 2011 | 2011 | 2011 | 747 | 0.37 | 74 |
| 13.81 | 7673 | 7673 | 6886 | 747 | 0.10 | -1 |
| 15.06 | 2007 | 2007 | 1492 | 744 | 0.37 | 76 |
| 14.62 | 7673 | 7673 | 6636 | 455 | 0.06 | 74 |
| 14.06 | 7676 | 7676 | 6053 | 388 | 0.05 | 78 |
| 12.75 | 7677 | 7677 | 7677 | 0 | 0.00 | -1 |
| 15.69 | 39 | 39 | 39 | 0 | 0.00 | 76 |

`extent_ok` is true on every row, so the mapping covers the sequence it should; what it contains is wrong, not how much of it there is. Sixty-three requests were live at each of these moments.

## What would settle it

Which path frees a slot that a request still holds. `SGLangKVPageLease` keeps prefill KV alive until the receiver's ACK releases the source, so a release that runs while a lease still references the page fits, and so does a release-then-reallocate on the decode side. Neither is established.

**Comment by jzheng17** [comment]
The two faults come from one place: two threads allocate from the same KV allocator without a shared lock, and the allocator's `alloc` is a non-atomic read-modify-write.

## The race

`TokenToKVPoolAllocator.alloc` in `sglang/srt/mem_cache/allocator/token.py`:

```python
select_index = self.free_pages[:need_size]
self.free_pages = self.free_pages[need_size:]
return select_index
```

Two callers that interleave between those two statements read the same `free_pages`, slice the same leading entries, and both receive them. The second assignment overwrites the first. There is no lock anywhere in `sglang/srt/mem_cache/allocator/` — `grep -rn "threading.Lock\|Lock()"` over that package returns nothing, which is fine while one thread owns the allocator.

On a PD Decode half two threads own it.

- **Scheduler thread.** `stage/runtime.py` starts the scheduler on its own `threading.Thread`, and its decode steps allocate from `self.token_to_kv_pool_allocator`.
- **Comm thread.** `_run_rank_control` is a coroutine on the comm engine's event loop. A `KVTransferPrepareMessage` there calls `prepare_kv_receive` → `receiver.reserve` → `self._allocator.alloc(count)`.

They are the same allocator: `bind_pd_runtime` constructs `AllocatorKVReceiver(token_to_kv_pool_allocator=self.token_to_kv_pool_allocator, ...)`.

`AllocatorKVReceiver` does hold `self._lock` across `reserve`, but that lock protects the adapter's own reservation dictionaries. The scheduler thread never takes it.

## It produces exactly the two signatures

Two allocations that return overlapping slots land either in two different requests' rows, which reads as `xshare > 0`, or in one request's row when a reserve and a decode-step allocation for the same request collide, which reads as `uniq_ok` false. Both were measured, 9 of 11 degraded rows against 2 of 75 controls within prompt class.

It also accounts for the rest of the shape:

- **PD only.** A colocated replica has no second thread reserving from the allocator.
- **Rises with queue depth.** More concurrent decode-step allocations widen the chance of landing inside the window.
- **Needs mixed prompt lengths.** Uniform long prompts did not reproduce it in six runs; `--pflags 42,1150,4400` did, twice. Varied `need_size` keeps the free list churning.
- **Low absolute rate.** The window is a few bytecodes, and the measured rate is 0.07% to 0.28%.
- **`extent_ok` true on every degraded row.** The mapping covers the right span; what it contains is wrong.
- **`reuse_gap` 73 to 78 on the duplicated rows.** Those slots had just been in use.

## What is verified and what is inferred

Verified by reading the code on this revision: the allocator has no lock, `alloc` is a non-atomic read-modify-write, `reserve` runs on the comm event loop, the scheduler runs on its own thread, and both use the same allocator instance.

Inferred: that this race is what produced the measured rows. The signatures match and no other path in the Decode half allocates from that allocator, but a direct capture of two overlapping `alloc` returns has not been taken.

## Fix

Serialize the two callers. The narrowest change is a lock the allocator's users share on this path — taken by `AllocatorKVReceiver.reserve` around its `alloc`, and by the scheduler around its own allocation — or hand the reservation to the scheduler thread instead of allocating on the comm thread. The second removes the sharing rather than guarding it, and the ready message already round-trips, so the reserve does not have to complete inline.

@MelodyyyYin this sits in the receiver you added in PR 3, so flagging it rather than picking the fix: whether `reserve` should allocate at all, or hand the request to the scheduler thread and answer once it has, is a design call on that layer.

**Comment by jzheng17** [comment]
Direct observation, not inference. Instrumenting `TokenToKVPoolAllocator.alloc` to record which thread received which slots, and `free` to release them, one 3,000-request run logged **21,309 cross-thread overlaps** — a slot handed to one thread while another still held it — alongside 11 empty completions.

```
ALLOCRACE slot=523027 again_by=scheduler-thinker_decode seq=714 prev_by=MainThread prev_seq=713 n=42
```

Both directions occur, which is what a lockless read-modify-write predicts:

| Thread that received a slot still held elsewhere | Count |
| --- | --- |
| `scheduler-thinker_decode`, the scheduler thread | 13,166 |
| `MainThread`, the comm event loop | 8,143 |

`MainThread` is where `_run_rank_control` runs `prepare_kv_receive` → `reserve` → `alloc`. `scheduler-thinker_decode` is where decode steps allocate. They share one allocator, which `bind_pd_runtime` hands to `AllocatorKVReceiver` directly.

A first version of this probe recorded allocations without releasing on `free`, so ordinary reuse after a free counted as an overlap. Its numbers are discarded; the run above releases correctly.

This closes the question the earlier comment left open. What remains is the fix, and that is a design call on the receiver: guard the allocator on this path with a lock both callers take, or stop allocating on the comm thread and hand the reservation to the scheduler thread, answering once it has. The ready message already round-trips, so the reserve does not have to complete inline.
